posthog/caching/insights_api.py

```diff
@@ -1,46 +1,112 @@
 from datetime import datetime, timedelta
 from math import ceil
+from time import sleep
 from typing import Optional, Tuple, Union
+from rest_framework import request
 
 import pytz
 
-from posthog.caching.calculate_results import calculate_cache_key
+from posthog.caching.calculate_results import CLICKHOUSE_MAX_EXECUTION_TIME, calculate_cache_key
 from posthog.caching.insight_caching_state import InsightCachingState
 from posthog.models import DashboardTile, Insight
 from posthog.models.filters.utils import get_filter
+from posthog.utils import refresh_requested_by_client
 
 """
 Utilities used by the insights API to determine whether
 or not to refresh an insight upon a client request to do so
 """
 
-# default minimum wait time for refreshing an insight
-DEFAULT_CLIENT_INSIGHT_ALLOWED_REFRESH_FREQUENCY = timedelta(minutes=15)
+# Default minimum wait time for refreshing an insight
+BASE_MINIMUM_INSIGHT_REFRESH_INTERVAL = timedelta(minutes=15)
+# Wait time for short-term insights
+REDUCED_MINIMUM_INSIGHT_REFRESH_INTERVAL = timedelta(minutes=3)
+# Wait time for insights on shared insight/dashboard pages
+INCREASED_MINIMUM_INSIGHT_REFRESH_INTERVAL = timedelta(minutes=30)
 
 
-# returns should_refresh, refresh_frequency
-def should_refresh_insight(insight: Insight, dashboard_tile: Optional[DashboardTile]) -> Tuple[bool, timedelta]:
+def should_refresh_insight(
+    insight: Insight, dashboard_tile: Optional[DashboardTile], *, request: request.Request, is_shared=False
```

Comment:
I'd be tempted to pass in the result of `refresh_requested_by_client(request)` instead of the request. Just to avoid passing the request around but only a nit